django/db/models/base.py

```diff
@@ -203,26 +203,35 @@ def __new__(cls, name, bases, attrs):
                 if isinstance(field, OneToOneField):
                     related = resolve_relation(new_class, field.remote_field.model)
                     parent_links[make_model_tuple(related)] = field
+
+        # Track fields inherited from base models.
+        inherited_attributes = set()
         # Do the appropriate setup for any model parents.
-        for base in parents:
+        for base in new_class.mro():
             original_base = base
-            if not hasattr(base, '_meta'):
+            if base not in parents or not hasattr(base, '_meta'):
                 # Things without _meta aren't functional models, so they're
                 # uninteresting parents.
+                inherited_attributes |= base.__dict__.keys()
```

Comment:
I guess this needs to be something like `inherited_attributes |= set(base.__dict__.keys())` to work on Python 2.